src/core/DY.Core.Bytes.fst

```diff
@@ -91,6 +91,117 @@ let rec length b =
+  | Concat left right ->
+    bytes_well_formed tr left /\
+    bytes_well_formed tr right
+  | AeadEnc key nonce msg ad ->
+    bytes_well_formed tr key /\
+    bytes_well_formed tr nonce /\
+    bytes_well_formed tr msg /\
+    bytes_well_formed tr ad
+  | Pk sk ->
+    bytes_well_formed tr sk
+  | PkEnc pk nonce msg ->
+    bytes_well_formed tr pk /\
+    bytes_well_formed tr nonce /\
+    bytes_well_formed tr msg
+  | Vk sk ->
+    bytes_well_formed tr sk
+  | Sign sk nonce msg ->
+    bytes_well_formed tr sk /\
+    bytes_well_formed tr nonce /\
+    bytes_well_formed tr msg
+  | Hash msg ->
+    bytes_well_formed tr msg
+  | DhPub sk ->
+    bytes_well_formed tr sk
+  | Dh sk pk ->
+    bytes_well_formed tr sk /\
+    bytes_well_formed tr pk
+  | KdfExtract salt ikm ->
+    bytes_well_formed tr salt /\
+    bytes_well_formed tr ikm
+  | KdfExpand prk info len ->
+    bytes_well_formed tr prk /\
+    bytes_well_formed tr info
+  | KemPub sk ->
+    bytes_well_formed tr sk
+  | KemEncap pk nonce ->
+    bytes_well_formed tr pk /\
+    bytes_well_formed tr nonce
+  | KemSecretShared nonce ->
+    bytes_well_formed tr nonce
+
+val bytes_well_formed_later:
+  tr1:trace -> tr2:trace ->
+  b:bytes ->
+  Lemma
+  (requires
+    bytes_well_formed tr1 b /\
+    tr1 <$ tr2
+  )
+  (ensures bytes_well_formed tr2 b)
+  [SMTPat (bytes_well_formed tr1 b);
+   SMTPat (bytes_well_formed tr2 b);
+   SMTPat (tr1 <$ tr2)]
+let rec bytes_well_formed_later tr1 tr2 b =
+  normalize_term_spec bytes_well_formed;
+  introduce forall b_sub. (b_sub << b /\ bytes_well_formed tr1 b_sub) ==> bytes_well_formed tr2 b_sub with (
+    introduce _ ==> _ with _. (
+      bytes_well_formed_later tr1 tr2 b_sub
+    )
+  );
+  match b with
+  | Literal buf -> ()
+  | Rand usg len time -> (
+    assert(event_at tr1 time (get_event_at tr1 time))
+  )
+  | Concat left right -> ()
+  | AeadEnc key nonce msg ad -> ()
+  | Pk sk -> ()
+  | PkEnc pk nonce msg -> ()
+  | Vk sk -> ()
+  | Sign sk nonce msg -> ()
+  | Hash msg -> ()
+  | DhPub sk -> ()
+  | Dh sk pk -> ()
+  | KdfExtract salt ikm -> ()
+  | KdfExpand prk info len -> ()
+  | KemPub sk -> ()
+  | KemEncap pk nonce -> ()
+  | KemSecretShared nonce -> ()
+
+/// Many lemmas with SMT patterns can help proving `bytes_well_formed`.
+/// These lemmas are not expected to be useful in most situations:
+/// - in general we have `bytes_invariant` that implies `bytes_well_formed`
+///   (see theorem `bytes_invariant_implies_well_formed`),
+/// - the only place where we cannot rely on `bytes_invariant`
+///   are in the `pred_later` lemmas inside `crypto_predicates`,
+///   which is only a small amount of the security proof.
+/// Hence the SMT patterns do not trigger by default,
+/// and can be enabled by doing `assert (enable_bytes_well_formed_smtpats tr);`.
+/// See https://github.com/FStarLang/FStar/wiki/Quantifiers-and-patterns
+/// for more information on this technique.
+
+let enable_bytes_well_formed_smtpats (tr:trace) = True
```

Comment:
Nit: This seems like it might be a bit hard to discover when it's around 200 lines into a large file. I think it's fine for now, but if we're going to start having more of these flags, we should consider if there's a clearer way to (a) make it easy to tell what flags exist and (b) make it easy to tell what lemmas a given flag enables. Of these, I think (b) is already pretty well handled (as we can just search for the flag name), and likewise, if we have a lemma that we know we want to use, it is easy to find the corresponding flag, but it might be nice for useability long-term to have a nice list somewhere of flags paired with a brief description of what kinds of lemmas they relate to (like the comment included here).

Reply:
Opened #64 to track this.